diff --git a/Modules/_io/iobase.c b/Modules/_io/iobase.c
--- a/Modules/_io/iobase.c
+++ b/Modules/_io/iobase.c
@@ -208,7 +208,8 @@ int
        resurrect the object as calling close() can invoke arbitrary code. */
     is_zombie = (Py_REFCNT(self) == 0);
     if (is_zombie) {
-        ++Py_REFCNT(self);
+        /* ++Py_REFCNT(self); */
+        Py_INCREF(self);
     }
     PyErr_Fetch(&tp, &v, &tb);
     /* If `closed` doesn't exist or can't be evaluated as bool, then the
