diff --git a/Python/ceval.c b/Python/ceval.c
index 2b16191..b73ab62 100644
--- a/Python/ceval.c
+++ b/Python/ceval.c
@@ -70,6 +70,10 @@ ppc_getcounter(uint64 *v)
                          "=a" (((int*)&(val))[0]), "=d" (((int*)&(val))[1]));
 
 
+#elif defined(WITH_INTRINSICS)
+
+#define READ_TIMESTAMP(var) (var = _Py_rdtsc())
+
 #else
 
 #error "Don't know how to implement timestamp counter for this architecture"
@@ -297,12 +301,15 @@ static int pending_async_exc = 0;
 int
 PyEval_ThreadsInitialized(void)
 {
+    Px_RETURN(0);
     return gil_created();
 }
 
 void
 PyEval_InitThreads(void)
 {
+    if (_PyParallel_GetActiveContext())
+        return;
     if (gil_created())
         return;
     create_gil();
@@ -315,6 +322,8 @@ PyEval_InitThreads(void)
 void
 _PyEval_FiniThreads(void)
 {
+    if (_PyParallel_GetActiveContext())
+        return;
     if (!gil_created())
         return;
     destroy_gil();
@@ -324,15 +333,21 @@ _PyEval_FiniThreads(void)
 void
 PyEval_AcquireLock(void)
 {
-    PyThreadState *tstate = PyThreadState_GET();
+    PyThreadState *tstate;
+    if (_PyParallel_GetActiveContext())
+        return;
+    tstate = PyThreadState_GET();
     if (tstate == NULL)
         Py_FatalError("PyEval_AcquireLock: current thread state is NULL");
+    Py_GUARD_AGAINST_PX_ONLY();
     take_gil(tstate);
 }
 
 void
 PyEval_ReleaseLock(void)
 {
+    if (_PyParallel_GetActiveContext())
+        return;
     /* This function must succeed when the current thread state is NULL.
        We therefore avoid PyThreadState_GET() which dumps a fatal error
        in debug mode.
@@ -344,6 +359,8 @@ PyEval_ReleaseLock(void)
 void
 PyEval_AcquireThread(PyThreadState *tstate)
 {
+    if (_PyParallel_GetActiveContext())
+        return;
     if (tstate == NULL)
         Py_FatalError("PyEval_AcquireThread: NULL new thread state");
     /* Check someone has called PyEval_InitThreads() to create the lock */
@@ -357,6 +374,8 @@ PyEval_AcquireThread(PyThreadState *tstate)
 void
 PyEval_ReleaseThread(PyThreadState *tstate)
 {
+    if (_PyParallel_GetActiveContext())
+        return;
     if (tstate == NULL)
         Py_FatalError("PyEval_ReleaseThread: NULL thread state");
     if (PyThreadState_Swap(NULL) != tstate)
@@ -374,7 +393,11 @@ PyEval_ReInitThreads(void)
 {
     _Py_IDENTIFIER(_after_fork);
     PyObject *threading, *result;
-    PyThreadState *tstate = PyThreadState_GET();
+    PyThreadState *tstate;
+    if (_PyParallel_GetActiveContext())
+        return;
+
+    tstate = PyThreadState_GET();
 
     if (!gil_created())
         return;
@@ -412,6 +435,7 @@ static int pending_async_exc = 0;
 void
 _PyEval_SignalAsyncExc(void)
 {
+    Px_VOID();
     SIGNAL_ASYNC_EXC();
 }
 
@@ -422,7 +446,10 @@ _PyEval_SignalAsyncExc(void)
 PyThreadState *
 PyEval_SaveThread(void)
 {
-    PyThreadState *tstate = PyThreadState_Swap(NULL);
+    PyThreadState *tstate;
+    if (_PyParallel_GetActiveContext())
+        return NULL;
+    tstate = PyThreadState_Swap(NULL);
     if (tstate == NULL)
         Py_FatalError("PyEval_SaveThread: NULL tstate");
 #ifdef WITH_THREAD
@@ -435,6 +462,8 @@ PyEval_SaveThread(void)
 void
 PyEval_RestoreThread(PyThreadState *tstate)
 {
+    if (_PyParallel_GetActiveContext())
+        return;
     if (tstate == NULL)
         Py_FatalError("PyEval_RestoreThread: NULL tstate");
 #ifdef WITH_THREAD
@@ -651,6 +679,7 @@ int
 Py_MakePendingCalls(void)
 {
     static int busy = 0;
+    Py_GUARD();
     if (busy)
         return 0;
     busy = 1;
@@ -689,12 +718,14 @@ int _Py_CheckRecursionLimit = Py_DEFAULT_RECURSION_LIMIT;
 int
 Py_GetRecursionLimit(void)
 {
+    Py_GUARD();
     return recursion_limit;
 }
 
 void
 Py_SetRecursionLimit(int new_limit)
 {
+    Py_GUARD();
     recursion_limit = new_limit;
     _Py_CheckRecursionLimit = recursion_limit;
 }
@@ -1236,6 +1267,15 @@ PyEval_EvalFrameEx(PyFrameObject *f, int throwflag)
         assert(stack_pointer >= f->f_valuestack); /* else underflow */
         assert(STACK_LEVEL() <= co->co_stacksize);  /* else overflow */
 
+#ifdef WITH_PARALLEL
+        /* Parallel threads skip the normal thread periodic tasks below. */
+
+        if (tstate->is_parallel_thread) {
+            goto fast_next_opcode_notracing;
+        }
+#endif /* WITH_PARALLEL */
+
+
         /* Do periodic things.  Doing this every time through
            the loop would add too much overhead, so we do it
            only every Nth instruction.  We also do it if
@@ -1314,7 +1354,9 @@ PyEval_EvalFrameEx(PyFrameObject *f, int throwflag)
         }
 
         /* Extract opcode and argument */
-
+#ifdef WITH_PARALLEL
+    fast_next_opcode_notracing:
+#endif
         opcode = NEXTOP();
         oparg = 0;   /* allows oparg to be stored in a register because
             it doesn't have to be remembered across a full loop */
@@ -2295,9 +2337,27 @@ PyEval_EvalFrameEx(PyFrameObject *f, int throwflag)
             PREDICT(POP_JUMP_IF_TRUE);
             DISPATCH();
 
+#ifdef WITH_PARALLEL
+#define PREVENT_PARALLEL_IMPORT() do {                                      \
+    if (Py_PXCTX()) {                                                       \
+        PyErr_SetString(PyExc_ImportError,                                  \
+                        "import not permitted within parallel context");    \
+        break;                                                              \
+    }                                                                       \
+} while (0)
+#else
+#define PREVENT_PARALLEL_IMPORT()
+#endif
+
         TARGET(IMPORT_NAME)
         {
             _Py_IDENTIFIER(__import__);
+            PyObject *old_f_globals = NULL;
+            if (Py_PXCTX()) {
+                old_f_globals = f->f_globals;
+                f->f_globals = f->f_locals;
+            }
+            //PREVENT_PARALLEL_IMPORT();
             w = GETITEM(names, oparg);
             x = _PyDict_GetItemId(f->f_builtins, &PyId___import__);
             if (x == NULL) {
@@ -2338,11 +2398,14 @@ PyEval_EvalFrameEx(PyFrameObject *f, int throwflag)
             READ_TIMESTAMP(intr1);
             Py_DECREF(w);
             SET_TOP(x);
+            if (old_f_globals)
+                f->f_globals = old_f_globals;
             if (x != NULL) DISPATCH();
             break;
         }
 
         TARGET(IMPORT_STAR)
+            PREVENT_PARALLEL_IMPORT();
             v = POP();
             PyFrame_FastToLocals(f);
             if ((x = f->f_locals) == NULL) {
@@ -2359,6 +2422,7 @@ PyEval_EvalFrameEx(PyFrameObject *f, int throwflag)
             break;
 
         TARGET(IMPORT_FROM)
+            PREVENT_PARALLEL_IMPORT();
             w = GETITEM(names, oparg);
             v = TOP();
             READ_TIMESTAMP(intr0);
