diff --git a/Python/sysmodule.c b/Python/sysmodule.c
index 2f700e6..361a470 100644
--- a/Python/sysmodule.c
+++ b/Python/sysmodule.c
@@ -1606,8 +1606,11 @@ _PySys_Init(void)
     SET_SYS_FROM_STRING("hexversion",
                          PyLong_FromLong(PY_VERSION_HEX));
     SET_SYS_FROM_STRING("_mercurial",
-                        Py_BuildValue("(szz)", "CPython", _Py_hgidentifier(),
-                                      _Py_hgversion()));
+                        Py_BuildValue("(szz)", "CPython", _Py_identifier(),
+                                      _Py_version()));
+    SET_SYS_FROM_STRING("_git",
+                        Py_BuildValue("(szz)", "CPython", _Py_identifier(),
+                                      _Py_version()));
     SET_SYS_FROM_STRING("dont_write_bytecode",
                          PyBool_FromLong(Py_DontWriteBytecodeFlag));
     SET_SYS_FROM_STRING("api_version",
