diff --git a/Objects/sliceobject.c b/Objects/sliceobject.c
index 1593335..532ef69 100644
--- a/Objects/sliceobject.c
+++ b/Objects/sliceobject.c
@@ -89,6 +89,8 @@ static PySliceObject *slice_cache = NULL;
 void PySlice_Fini(void)
 {
     PySliceObject *obj = slice_cache;
+    if (Py_PXCTX())
+        return;
     if (obj != NULL) {
         slice_cache = NULL;
         PyObject_Del(obj);
@@ -103,7 +105,8 @@ PyObject *
 PySlice_New(PyObject *start, PyObject *stop, PyObject *step)
 {
     PySliceObject *obj;
-    if (slice_cache != NULL) {
+
+    if (!Py_PXCTX() && slice_cache != NULL) {
         obj = slice_cache;
         slice_cache = NULL;
         _Py_NewReference((PyObject *)obj);
@@ -277,6 +280,9 @@ Create a slice object.  This is used for extended slicing (e.g. a[0:10:2]).");
 static void
 slice_dealloc(PySliceObject *r)
 {
+    if (Py_PXCTX())
+        return;
+
     Py_DECREF(r->step);
     Py_DECREF(r->start);
     Py_DECREF(r->stop);
