diff --git a/Modules/main.c b/Modules/main.c
index 3cfc9c3..fe4befa 100644
--- a/Modules/main.c
+++ b/Modules/main.c
@@ -337,6 +337,10 @@ Py_Main(int argc, wchar_t **argv)
     orig_argc = argc;           /* For Py_GetArgcArgv() */
     orig_argv = argv;
 
+#ifdef WITH_PARALLEL
+    _PyParallel_Init();
+#endif
+
     /* Hash randomization needed early for all string operations
        (including -W and -X options). */
     _PyOS_opterr = 0;  /* prevent printing the error in 1st pass */
@@ -475,7 +479,7 @@ Py_Main(int argc, wchar_t **argv)
         return usage(0, argv[0]);
 
     if (version) {
-        fprintf(stderr, "Python %s\n", PY_VERSION);
+        fprintf(stderr, "PyParallel %s\n", PY_VERSION);
         return 0;
     }
 
@@ -650,7 +654,7 @@ Py_Main(int argc, wchar_t **argv)
     if (!Py_QuietFlag && (Py_VerboseFlag ||
                         (command == NULL && filename == NULL &&
                          module == NULL && stdin_is_interactive))) {
-        fprintf(stderr, "Python %s on %s\n",
+        fprintf(stderr, "PyParallel %s on %s\n",
             Py_GetVersion(), Py_GetPlatform());
         if (!Py_NoSiteFlag)
             fprintf(stderr, "%s\n", COPYRIGHT);
