diff --git a/PCbuild/build_ssl.py b/PCbuild/build_ssl.py
index 5a7a89e..17f0fba 100644
--- a/PCbuild/build_ssl.py
+++ b/PCbuild/build_ssl.py
@@ -170,7 +170,12 @@ def main():
         make_flags = "-a"
     # perl should be on the path, but we also look in "\perl" and "c:\\perl"
     # as "well known" locations
-    perls = find_all_on_path("perl.exe", ["\\perl\\bin", "C:\\perl\\bin"])
+    perls = find_all_on_path("perl.exe", [
+        "\\perl\\bin",
+        "C:\\perl\\bin",
+        "\\perl64\\bin",
+        "C:\\perl64\\bin",
+    ])
     perl = find_working_perl(perls)
     if perl:
         print("Found a working perl at '%s'" % (perl,))
