diff --git a/PCbuild/pyproject.props b/PCbuild/pyproject.props
index 8155231..9feeb71 100644
--- a/PCbuild/pyproject.props
+++ b/PCbuild/pyproject.props
@@ -17,7 +17,7 @@
     <PythonExe>$(OutDir)python$(PyDebugExt).exe</PythonExe>
     <KillPythonExe>$(OutDir)kill_python$(PyDebugExt).exe</KillPythonExe>
     <externalsDir>..\..</externalsDir>
-    <sqlite3Dir>$(externalsDir)\sqlite-3.7.12</sqlite3Dir>
+    <sqlite3Dir>..\contrib\sqlite3</sqlite3Dir>
     <bz2Dir>$(externalsDir)\bzip2-1.0.6</bz2Dir>
     <lzmaDir>$(externalsDir)\xz-5.0.3</lzmaDir>
     <opensslDir>$(externalsDir)\openssl-1.0.1e</opensslDir>
@@ -27,13 +27,20 @@
     <tcltkLibDebug>$(tcltkDir)\lib\tcl85g.lib;$(tcltkDir)\lib\tk85g.lib</tcltkLibDebug>
     <tcltk64Lib>$(tcltk64Dir)\lib\tcl85.lib;$(tcltk64Dir)\lib\tk85.lib</tcltk64Lib>
     <tcltk64LibDebug>$(tcltk64Dir)\lib\tcl85g.lib;$(tcltk64Dir)\lib\tk85g.lib</tcltk64LibDebug>
+    <Win81SDKDir>C:\Program Files (x86)\Windows Kits\8.1</Win81SDKDir>
+    <Win81SDKIncludeDir>$(Win81SDKDir)\Include</Win81SDKIncludeDir>
+    <Win10SDKVersion>10.0.10240.0</Win10SDKVersion>
+    <Win10SDKDir>C:\Program Files (x86)\Windows Kits\10</Win10SDKDir>
+    <Win10SDKIncludeDir>$(Win10SDKDir)\Include\$(Win10SDKVersion)</Win10SDKIncludeDir>
+    <WinSDKLibDir>$(Win10SDKDir)\Lib\$(Win10SDKVersion)</WinSDKLibDir>
+    <WinSDKIncludeDir>$(Win10SDKIncludeDir)</WinSDKIncludeDir>
   </PropertyGroup>
   <ItemDefinitionGroup>
     <ClCompile>
       <Optimization>MaxSpeed</Optimization>
       <InlineFunctionExpansion>OnlyExplicitInline</InlineFunctionExpansion>
       <IntrinsicFunctions>true</IntrinsicFunctions>
-      <AdditionalIncludeDirectories>..\Include; ..\PC;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+      <AdditionalIncludeDirectories>$(WinSDKIncludeDir)\shared;$(WinSDKIncludeDir)\km;$(WinSDKIncludeDir)\um; ..\Include; ..\PC;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
       <PreprocessorDefinitions>_WIN32;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <StringPooling>true</StringPooling>
       <ExceptionHandling>
