diff --git a/PCbuild/pythoncore.vcxproj b/PCbuild/pythoncore.vcxproj
index 8b5c42f..cbe80d2 100644
--- a/PCbuild/pythoncore.vcxproj
+++ b/PCbuild/pythoncore.vcxproj
@@ -123,6 +123,7 @@
     <Import Project="pyproject.props" />
     <Import Project="x64.props" />
   </ImportGroup>
+  <Import Project="cython.props" />
   <PropertyGroup Label="UserMacros" />
   <PropertyGroup>
     <_ProjectFileVersion>10.0.30319.1</_ProjectFileVersion>
@@ -158,6 +159,7 @@
     <TargetName Condition="'$(Configuration)|$(Platform)'=='PGInstrument|x64'">$(PyDllName)</TargetName>
     <TargetName Condition="'$(Configuration)|$(Platform)'=='PGUpdate|x64'">$(PyDllName)</TargetName>
     <TargetName Condition="'$(Configuration)|$(Platform)'=='Release|x64'">$(PyDllName)</TargetName>
+    <PreBuildEventUseInBuild Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">false</PreBuildEventUseInBuild>
   </PropertyGroup>
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
     <ClCompile>
@@ -165,6 +167,8 @@
       <AdditionalIncludeDirectories>..\Python;..\Modules\zlib;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
       <PreprocessorDefinitions>_USRDLL;Py_BUILD_CORE;Py_ENABLE_SHARED;WIN32;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <RuntimeLibrary>MultiThreadedDLL</RuntimeLibrary>
+      <BrowseInformation>true</BrowseInformation>
+      <AssemblerOutput>All</AssemblerOutput>
     </ClCompile>
     <ResourceCompile>
       <PreprocessorDefinitions>NDEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
@@ -201,6 +205,8 @@ IF %ERRORLEVEL% NEQ 0 (
       <AdditionalIncludeDirectories>..\Python;..\Modules\zlib;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
       <PreprocessorDefinitions>_USRDLL;Py_BUILD_CORE;Py_ENABLE_SHARED;WIN32;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <RuntimeLibrary>MultiThreadedDLL</RuntimeLibrary>
+      <BrowseInformation>true</BrowseInformation>
+      <AssemblerOutput>All</AssemblerOutput>
     </ClCompile>
     <ResourceCompile>
       <PreprocessorDefinitions>NDEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
@@ -212,6 +218,7 @@ IF %ERRORLEVEL% NEQ 0 (
       <Command>"$(SolutionDir)make_buildinfo.exe" Release "$(IntDir)"</Command>
     </PreLinkEvent>
     <Link>
+      <AdditionalLibraryDirectories>$(WinSDKLibDir)\km\x64;$(WinSDKLibDir)\um\x64</AdditionalLibraryDirectories>
       <AdditionalDependencies>$(IntDir)getbuildinfo.o;%(AdditionalDependencies)</AdditionalDependencies>
       <IgnoreSpecificDefaultLibraries>libc;%(IgnoreSpecificDefaultLibraries)</IgnoreSpecificDefaultLibraries>
       <BaseAddress>0x1e000000</BaseAddress>
@@ -236,6 +243,8 @@ IF %ERRORLEVEL% NEQ 0 (
       <AdditionalIncludeDirectories>..\Python;..\Modules\zlib;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
       <PreprocessorDefinitions>_USRDLL;Py_BUILD_CORE;Py_ENABLE_SHARED;WIN32;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <RuntimeLibrary>MultiThreadedDebugDLL</RuntimeLibrary>
+      <BrowseInformation>true</BrowseInformation>
+      <AssemblerOutput>All</AssemblerOutput>
     </ClCompile>
     <ResourceCompile>
       <PreprocessorDefinitions>_DEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
@@ -274,6 +283,8 @@ IF %ERRORLEVEL% NEQ 0 (
       <AdditionalIncludeDirectories>..\Python;..\Modules\zlib;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
       <PreprocessorDefinitions>_USRDLL;Py_BUILD_CORE;Py_ENABLE_SHARED;WIN32;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <RuntimeLibrary>MultiThreadedDebugDLL</RuntimeLibrary>
+      <BrowseInformation>true</BrowseInformation>
+      <AssemblerOutput>All</AssemblerOutput>
     </ClCompile>
     <ResourceCompile>
       <PreprocessorDefinitions>_DEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
@@ -285,6 +296,7 @@ IF %ERRORLEVEL% NEQ 0 (
       <Command>"$(SolutionDir)make_buildinfo.exe" Debug "$(IntDir)"</Command>
     </PreLinkEvent>
     <Link>
+      <AdditionalLibraryDirectories>$(WinSDKLibDir)\km\x64;$(WinSDKLibDir)\um\x64</AdditionalLibraryDirectories>
       <AdditionalDependencies>$(IntDir)getbuildinfo.o;%(AdditionalDependencies)</AdditionalDependencies>
       <IgnoreSpecificDefaultLibraries>libc;%(IgnoreSpecificDefaultLibraries)</IgnoreSpecificDefaultLibraries>
       <BaseAddress>0x1e000000</BaseAddress>
@@ -306,6 +318,8 @@ IF %ERRORLEVEL% NEQ 0 (
       <AdditionalIncludeDirectories>..\Python;..\Modules\zlib;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
       <PreprocessorDefinitions>_USRDLL;Py_BUILD_CORE;Py_ENABLE_SHARED;WIN32;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <RuntimeLibrary>MultiThreadedDLL</RuntimeLibrary>
+      <BrowseInformation>true</BrowseInformation>
+      <AssemblerOutput>All</AssemblerOutput>
     </ClCompile>
     <ResourceCompile>
       <PreprocessorDefinitions>NDEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
@@ -342,6 +356,8 @@ IF %ERRORLEVEL% NEQ 0 (
       <AdditionalIncludeDirectories>..\Python;..\Modules\zlib;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
       <PreprocessorDefinitions>_USRDLL;Py_BUILD_CORE;Py_ENABLE_SHARED;WIN32;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <RuntimeLibrary>MultiThreadedDLL</RuntimeLibrary>
+      <BrowseInformation>true</BrowseInformation>
+      <AssemblerOutput>All</AssemblerOutput>
     </ClCompile>
     <ResourceCompile>
       <PreprocessorDefinitions>NDEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
@@ -353,6 +369,7 @@ IF %ERRORLEVEL% NEQ 0 (
       <Command>"$(SolutionDir)make_buildinfo.exe" Release "$(IntDir)"</Command>
     </PreLinkEvent>
     <Link>
+      <AdditionalLibraryDirectories>$(WinSDKLibDir)\km\x64;$(WinSDKLibDir)\um\x64</AdditionalLibraryDirectories>
       <AdditionalDependencies>$(IntDir)getbuildinfo.o;%(AdditionalDependencies)</AdditionalDependencies>
       <IgnoreSpecificDefaultLibraries>libc;%(IgnoreSpecificDefaultLibraries)</IgnoreSpecificDefaultLibraries>
       <BaseAddress>0x1e000000</BaseAddress>
@@ -375,6 +392,8 @@ IF %ERRORLEVEL% NEQ 0 (
       <AdditionalIncludeDirectories>..\Python;..\Modules\zlib;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
       <PreprocessorDefinitions>_USRDLL;Py_BUILD_CORE;Py_ENABLE_SHARED;WIN32;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <RuntimeLibrary>MultiThreadedDLL</RuntimeLibrary>
+      <BrowseInformation>true</BrowseInformation>
+      <AssemblerOutput>All</AssemblerOutput>
     </ClCompile>
     <ResourceCompile>
       <PreprocessorDefinitions>NDEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
@@ -411,6 +430,8 @@ IF %ERRORLEVEL% NEQ 0 (
       <AdditionalIncludeDirectories>..\Python;..\Modules\zlib;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
       <PreprocessorDefinitions>_USRDLL;Py_BUILD_CORE;Py_ENABLE_SHARED;WIN32;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <RuntimeLibrary>MultiThreadedDLL</RuntimeLibrary>
+      <BrowseInformation>true</BrowseInformation>
+      <AssemblerOutput>All</AssemblerOutput>
     </ClCompile>
     <ResourceCompile>
       <PreprocessorDefinitions>NDEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
@@ -422,6 +443,7 @@ IF %ERRORLEVEL% NEQ 0 (
       <Command>"$(SolutionDir)make_buildinfo.exe" Release "$(IntDir)"</Command>
     </PreLinkEvent>
     <Link>
+      <AdditionalLibraryDirectories>$(WinSDKLibDir)\km\x64;$(WinSDKLibDir)\um\x64</AdditionalLibraryDirectories>
       <AdditionalDependencies>$(IntDir)getbuildinfo.o;%(AdditionalDependencies)</AdditionalDependencies>
       <IgnoreSpecificDefaultLibraries>libc;%(IgnoreSpecificDefaultLibraries)</IgnoreSpecificDefaultLibraries>
       <BaseAddress>0x1e000000</BaseAddress>
@@ -439,6 +461,18 @@ IF %ERRORLEVEL% NEQ 0 (
     </PreBuildEvent>
   </ItemDefinitionGroup>
   <ItemGroup>
+    <CythonFile Include="..\contrib\windows\avrt.pyx" />
+    <CythonFile Include="..\contrib\windows\adv.pyx" />
+    <CythonFile Include="..\contrib\windows\user.pyx" />
+    <CythonFile Include="..\contrib\windows\rtl.pyx" />
+    <CythonFile Include="..\contrib\windows\processthreads.pyx" />
+    <CythonFile Include="..\contrib\windows\error.pyx" />
+    <CythonFile Include="..\contrib\windows\base.pyx" />
+  </ItemGroup>
+  <ItemGroup>
+    <CythonCppFile Include="..\contrib\windows\gdiplus.pyx" />
+  </ItemGroup>
+  <ItemGroup>
     <ClInclude Include="..\Include\abstract.h" />
     <ClInclude Include="..\Include\accu.h" />
     <ClInclude Include="..\Include\asdl.h" />
@@ -565,8 +599,36 @@ IF %ERRORLEVEL% NEQ 0 (
     <ClInclude Include="..\Python\condvar.h" />
     <ClInclude Include="..\Python\importdl.h" />
     <ClInclude Include="..\Python\thread_nt.h" />
+    <ClInclude Include="..\Include\picohttpparser.h" />
+    <ClInclude Include="..\Include\pyparallel.h" />
+    <ClInclude Include="..\Include\pyparallel_util.h" />
+    <ClInclude Include="..\Include\pyparallel_odbc.h" />
+    <ClInclude Include="..\Include\pyparallel_http.h" />
+    <ClInclude Include="..\Include\pxlist.h" />
+    <ClInclude Include="..\Include\pxtimerobject.h" />
+    <ClInclude Include="..\Include\pxthreadobject.h" />
+    <ClInclude Include="..\Include\statics.h" />
+    <ClInclude Include="..\Include\ultrajson.h" />
+    <ClInclude Include="..\Python\ujson_py_defines.h" />
+    <ClInclude Include="..\Python\ujson_version.h" />
+    <ClInclude Include="..\contrib\windows\processthreads.h" />
+    <ClInclude Include="..\contrib\windows\error.h" />
+    <ClInclude Include="..\contrib\windows\_gdiplus.h" />
+    <ClInclude Include="..\contrib\windows\_adv.h" />
+    <ClInclude Include="..\contrib\windows\_avrt.h" />
   </ItemGroup>
   <ItemGroup>
+    <ClCompile Include="..\contrib\windows\_loader.c" />
+    <ClCompile Include="..\contrib\windows\_gdiplus.cpp" />
+    <ClCompile Include="..\contrib\windows\_avrt.c" />
+    <ClCompile Include="..\contrib\windows\gdiplus.cpp" />
+    <ClCompile Include="..\contrib\windows\base.c" />
+    <ClCompile Include="..\contrib\windows\error.c" />
+    <ClCompile Include="..\contrib\windows\processthreads.c" />
+    <ClCompile Include="..\contrib\windows\rtl.c" />
+    <ClCompile Include="..\contrib\windows\user.c" />
+    <ClCompile Include="..\contrib\windows\adv.c" />
+    <ClCompile Include="..\contrib\windows\avrt.c" />
     <ClCompile Include="..\Modules\_bisectmodule.c" />
     <ClCompile Include="..\Modules\_codecsmodule.c" />
     <ClCompile Include="..\Modules\_collectionsmodule.c" />
@@ -674,6 +736,8 @@ IF %ERRORLEVEL% NEQ 0 (
     <ClCompile Include="..\Objects\unicodectype.c" />
     <ClCompile Include="..\Objects\unicodeobject.c" />
     <ClCompile Include="..\Objects\weakrefobject.c" />
+    <ClCompile Include="..\Objects\pxtimerobject.c" />
+    <ClCompile Include="..\Objects\pxthreadobject.c" />
     <ClCompile Include="..\Parser\acceler.c" />
     <ClCompile Include="..\Parser\bitset.c" />
     <ClCompile Include="..\Parser\firstsets.c" />
@@ -736,6 +800,16 @@ IF %ERRORLEVEL% NEQ 0 (
     <ClCompile Include="..\Python\sysmodule.c" />
     <ClCompile Include="..\Python\thread.c" />
     <ClCompile Include="..\Python\traceback.c" />
+    <ClCompile Include="..\Python\pyparallel_http.c" />
+    <ClCompile Include="..\Python\picohttpparser.c" />
+    <ClCompile Include="..\Python\pyparallel.c" />
+    <ClCompile Include="..\Python\pxlist.c" />
+    <ClCompile Include="..\Python\statics.c" />
+    <ClCompile Include="..\Python\JSONtoObj.c" />
+    <ClCompile Include="..\Python\objToJSON.c" />
+    <ClCompile Include="..\Python\ujson.c" />
+    <ClCompile Include="..\Python\ultrajsondec.c" />
+    <ClCompile Include="..\Python\ultrajsonenc.c" />
   </ItemGroup>
   <ItemGroup>
     <ResourceCompile Include="..\PC\python_nt.rc" />
