diff --git a/Include/bytearrayobject.h b/Include/bytearrayobject.h
index eccd44c..c602e91 100644
--- a/Include/bytearrayobject.h
+++ b/Include/bytearrayobject.h
@@ -44,6 +44,8 @@ PyAPI_FUNC(PyObject *) PyByteArray_FromStringAndSize(const char *, Py_ssize_t);
 PyAPI_FUNC(Py_ssize_t) PyByteArray_Size(PyObject *);
 PyAPI_FUNC(char *) PyByteArray_AsString(PyObject *);
 PyAPI_FUNC(int) PyByteArray_Resize(PyObject *, Py_ssize_t);
+PyAPI_FUNC(PyByteArrayObject *) PyByteArray_Copy(PyByteArrayObject *);
+PyAPI_FUNC(PyObject *) PyByteArray_CopyObject(PyObject *);
 
 /* Macros, trading safety for speed */
 #ifndef Py_LIMITED_API
