diff --git a/Include/ceval.h b/Include/ceval.h
index 6811367..45e7ec0 100644
--- a/Include/ceval.h
+++ b/Include/ceval.h
@@ -177,11 +177,11 @@ PyAPI_FUNC(unsigned long) _PyEval_GetSwitchInterval(void);
 #endif
 
 #define Py_BEGIN_ALLOW_THREADS { \
-                        PyThreadState *_save; \
-                        _save = PyEval_SaveThread();
-#define Py_BLOCK_THREADS        PyEval_RestoreThread(_save);
-#define Py_UNBLOCK_THREADS      _save = PyEval_SaveThread();
-#define Py_END_ALLOW_THREADS    PyEval_RestoreThread(_save); \
+                        PyThreadState *_save = NULL; \
+                        if (!_PyParallel_GetActiveContext()) _save = PyEval_SaveThread();
+#define Py_BLOCK_THREADS        (_PyParallel_GetActiveContext() ? 0 : PyEval_RestoreThread(_save));
+#define Py_UNBLOCK_THREADS      (_PyParallel_GetActiveContext() ? 0 : (_save = PyEval_SaveThread()));
+#define Py_END_ALLOW_THREADS    (_PyParallel_GetActiveContext() ? 0 : (_save ? PyEval_RestoreThread(_save) : 0)); \
                  }
 
 #else /* !WITH_THREAD */
