diff --git a/Include/datetime.h b/Include/datetime.h
index 06cbc4a..1c5ce67 100644
--- a/Include/datetime.h
+++ b/Include/datetime.h
@@ -181,6 +181,9 @@ typedef struct {
 
 #define PyDelta_Check(op) PyObject_TypeCheck(op, &PyDateTime_DeltaType)
 #define PyDelta_CheckExact(op) (Py_TYPE(op) == &PyDateTime_DeltaType)
+/* Needed for pyparallel. */
+int _PyDateTime_Check(PyObject *op);
+int _PyDelta_Check(PyObject *op);
 
 #define PyTZInfo_Check(op) PyObject_TypeCheck(op, &PyDateTime_TZInfoType)
 #define PyTZInfo_CheckExact(op) (Py_TYPE(op) == &PyDateTime_TZInfoType)
