diff --git a/Include/py_curses.h b/Include/py_curses.h
index f2c08f6..9b0cca1 100644
--- a/Include/py_curses.h
+++ b/Include/py_curses.h
@@ -103,8 +103,8 @@ static void **PyCurses_API;
 #endif
 
 /* general error messages */
-static char *catchall_ERR  = "curses function returned ERR";
-static char *catchall_NULL = "curses function returned NULL";
+static const char *catchall_ERR  = "curses function returned ERR";
+static const char *catchall_NULL = "curses function returned NULL";
 
 /* Function Prototype Macros - They are ugly but very, very useful. ;-)
 
