diff --git a/Include/Python.h b/Include/Python.h
index a78a721..a2df83d 100644
--- a/Include/Python.h
+++ b/Include/Python.h
@@ -52,6 +52,8 @@
 
 #include "pyatomic.h"
 
+#include "pyparallel.h"
+
 /* Debug-mode build with pymalloc implies PYMALLOC_DEBUG.
  *  PYMALLOC_DEBUG is in error if pymalloc is not in use.
  */
@@ -129,4 +131,9 @@
 #include "fileutils.h"
 #include "pyfpe.h"
 
+#ifdef WITH_PARALLEL
+PyAPI_DATA(PyTypeObject) PxTimer_Type;
+PyAPI_DATA(PyTypeObject) PxThread_Type;
+#endif
+
 #endif /* !Py_PYTHON_H */
