diff --git a/Include/pythonrun.h b/Include/pythonrun.h
index e8a582d..9fae818 100644
--- a/Include/pythonrun.h
+++ b/Include/pythonrun.h
@@ -200,6 +200,10 @@ PyAPI_FUNC(int) _PyFrame_Init(void);
 PyAPI_FUNC(void) _PyFloat_Init(void);
 PyAPI_FUNC(int) PyByteArray_Init(void);
 PyAPI_FUNC(void) _PyRandom_Init(void);
+#ifdef WITH_PARALLEL
+PyAPI_FUNC(PyObject *) _PyParallel_ModInit(void);
+PyAPI_FUNC(PyObject *) _PyAsync_ModInit(void);
+#endif
 #endif
 
 /* Various internal finalizers */
