diff --git a/Lib/importlib/_bootstrap.py b/Lib/importlib/_bootstrap.py
--- a/Lib/importlib/_bootstrap.py
+++ b/Lib/importlib/_bootstrap.py
@@ -403,7 +403,7 @@ due to the addition of new opcodes).
 
 """
 _RAW_MAGIC_NUMBER = 3230 | ord('\r') << 16 | ord('\n') << 24
-_MAGIC_BYTES = bytes(_RAW_MAGIC_NUMBER >> n & 0xff for n in range(0, 25, 8))
+_MAGIC_BYTES = bytes([_RAW_MAGIC_NUMBER >> n & 0xff for n in range(0, 25, 8)])
 
 _PYCACHE = '__pycache__'
 
