diff --git a/PC/pyconfig.h b/PC/pyconfig.h
--- a/PC/pyconfig.h
+++ b/PC/pyconfig.h
@@ -88,6 +88,11 @@ WIN32 is still required for the locale m
 #define USE_SOCKET
 #endif
 
+#if !defined(WITH_PARALLEL) && !defined(DISABLE_PARALLEL) 
+#define WITH_PARALLEL 1
+#define WITH_INTRINSICS 1
+#endif
+
 /* CE6 doesn't have strdup() but _strdup(). Assume the same for earlier versions. */
 #if defined(MS_WINCE)
 #  include <stdlib.h>
@@ -155,6 +160,14 @@ WIN32 is still required for the locale m
 #endif
 #endif /* MS_WIN64 */
 
+#ifdef WITH_PARALLEL
+#define Py_WINVER 0x0601
+#define Py_NTDDI NTDDI_WIN7
+#ifdef Py_DEBUG
+#define WITH_TSC 1
+#endif
+#else
+
 /* set the version macros for the windows headers */
 #ifdef MS_WINX64
 /* 64 bit only runs on XP or greater */
@@ -166,6 +179,8 @@ WIN32 is still required for the locale m
 #define Py_NTDDI NTDDI_WIN2KSP4
 #endif
 
+#endif /* WITH_PARALLEL */
+
 /* We only set these values when building Python - we don't want to force
    these values on extensions, as that will affect the prototypes and
    structures exposed in the Windows headers. Even when building Python, we
